### PR TITLE
OCPNODE-2400: Add Conditions field to the config.node object's status

### PR DESCRIPTION
**Pull request by sairameshv**
Based on the discussion [here](https://docs.google.com/document/d/1z25jsBi9lZVvQdZICG3QrA6j2C-8UIF2lxpN-b4HM2c/edit#heading=h.7tru0r8qo1wf) it is necessary to add a note about the deprecation of the Cgroupsv1 usage via the nodes.config object's status field.

Hence added the "Conditions" field to the config.node object's status.

/cc @harche @rphillips

### CODE REVIEW

**Comment by harche** [comment]
I was wondering instead of using `Conditions` directly, should have something like `NodeConditions` (like we [tried this once with workerLatencyProfiles](https://github.com/openshift/api/pull/1187/files#diff-7ffc4ab9c9775f324fe2d76ccd536e7a646e95fe8561ef91c5f519826d6a8411L47))?

/cc @deads2k @rphillips

**Comment by sairameshv** [comment]
> I was wondering instead of using `Conditions` directly, should have something like `NodeConditions` (like we [tried this once with workerLatencyProfiles](https://github.com/openshift/api/pull/1187/files#diff-7ffc4ab9c9775f324fe2d76ccd536e7a646e95fe8561ef91c5f519826d6a8411L47))?
> 
> /cc @deads2k @rphillips

We have the field `NodeStatus` already and the `Conditions` is a generic term. Do we have to still rename it to `NodeConditions` ?

**Comment by harche** [comment]
In case of worker latency profiles, we  thought of adding a latency profile specific status in `NodeStatus`. Since nodes.config object deals with various node related features, I thought maybe if we have a status dedicated for each feature it might a good place to keep everything organized.

In this case you are adding `NodeStatus` to show cgroup v1 deprecation notice, so many be a simple condition is good enough (although you can still have `CgroupDeprecationStatus`) but for more complicated features (e.g. WorkerLatencyProfiles) having a dedicated status for each feature makes sense to me.

**Comment by haircommander** [comment]
I think if we find utility in node conditions generally, having a generic one inside the node status makes sense. if we have a cgroup specific one, then we'd probably deprecate the field when cgroupv1 was finally dropped. We wouldn't need to if there are uses for the generic field

**Comment by harche** [comment]
> if we have a cgroup specific one, then we'd probably deprecate the field when cgroupv1 was finally dropped

This will unlikely to get deprecated. Not sure why would we want to deprecate this node status field that warns the user of a deprecated feature (cgroup v1). We would always want to warn user if they ever find themselves using cgroup v1, more so if the cgroup v1 is deprecated as a feature.

**Comment by haircommander** [comment]
fair enough. I still think generic conditions could be useful

**Comment by rphillips** [comment]
The config's node object is a singleton for the entire cluster. Does it make sense to add a condition here for N nodes? A better spot would be on the Node object itself.

**Comment by harche** [comment]
@rphillips OCP is deprecating cgroup v1 support, upstream k8s is just going to keep it in maintenance mode. This PR is to allow us display deprecation notice if the OCP cluster is set to use cgroup v1. Also in case of OCP, our cgroup version settings are cluster-wide (managed via config.node object), so it makes sense to update the status of the object driving the cgroup version change.

File: config/v1/types_node.go [hunk 1]
```diff
@@ -48,7 +48,12 @@ type NodeSpec struct {
 	WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`
 }
 
-type NodeStatus struct{}
+type NodeStatus struct {
+	// conditions represent the observations of the current state.
```
Comment (haircommander) [inline config/v1/types_node.go:52]: should the go doc have more information like that in the description in the crds below?

File: config/v1/types_node.go [hunk 2]
```diff
@@ -48,7 +48,12 @@ type NodeSpec struct {
...
+	// +listType=map
+	// +listMapKey=type
```
Comment (rphillips) [inline config/v1/types_node.go:54]: This might be missing // +optional

Comment (sairameshv) [reply config/v1/types_node.go:54]: Addressed

File: config/v1/types_node.go (hunk 1 above)
Comment (sairameshv) [reply config/v1/types_node.go]: Added a generic info about the conditions field.

File: config/v1/types_node.go [hunk 3]
```diff
@@ -48,7 +48,13 @@ type NodeSpec struct {
 	WorkerLatencyProfile WorkerLatencyProfileType `json:"workerLatencyProfile,omitempty"`
 }
 
-type NodeStatus struct{}
+type NodeStatus struct {
+	// Conditions contain the details and the current state of the nodes.config object
```
Comment (JoelSpeed) [inline config/v1/types_node.go:52]: Use the json tag version of the field name to start the comment.

Do you have any known condition types you could enumerate here?

File: config/v1/types_node.go [hunk 4]
```diff
@@ -48,7 +48,13 @@ type NodeSpec struct {
...
+	// +listType=map
+	// +listMapKey=type
```
Comment (JoelSpeed) [inline config/v1/types_node.go:54]: If you check other versions of this field, they have strategic merge patch keys as well, can you add those

Comment (sairameshv) [reply config/v1/types_node.go:56]: Addressed

File: config/v1/types_node.go (hunk 3 above)
Comment (sairameshv) [reply config/v1/types_node.go]: Addressed, There are no known condition types to enumerate as of now.

**Comment by sairameshv** [comment]
> @sairameshv Changes look ok, we are generally asking people to use feature gates for new features though, are you able to gate this addition? What's the implementation state on the operator side look like right now?

@JoelSpeed As mentioned in the description, we intend to add a note/message in the config node object's status field about the deprecation of the usage of cgroupsv1 mode if the cluster is using cgroupsv1. There is no functionality change and hence I don't think a featuregate is needed here. Do you think otherwise?

cc: @harche

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-upgrade-minor

Perma broken right now, am working on a fix

**Comment by sairameshv** [comment]
I don't know if this `hold` label gets removed by the bot, @JoelSpeed could you help in getting this merged?

**Pull request merged**